lite/dbprovider.go

```diff
@@ -0,0 +1,260 @@
+package lite
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	amino "github.com/tendermint/go-amino"
+	cryptoAmino "github.com/tendermint/tendermint/crypto/encoding/amino"
+	dbm "github.com/tendermint/tendermint/libs/db"
+	log "github.com/tendermint/tendermint/libs/log"
+	lerr "github.com/tendermint/tendermint/lite/errors"
+	"github.com/tendermint/tendermint/types"
+)
+
+type DBProvider struct {
+	logger log.Logger
+	label  string
+	db     dbm.DB
+	cdc    *amino.Codec
+	limit  int
+}
+
+func NewDBProvider(label string, db dbm.DB) *DBProvider {
+	//db = dbm.NewDebugDB("db provider "+cmn.RandStr(4), db)
+	cdc := amino.NewCodec()
+	cryptoAmino.RegisterAmino(cdc)
+	dbp := &DBProvider{
+		logger: log.NewNopLogger(),
+		label:  label,
+		db:     db,
+		cdc:    cdc,
+	}
+	return dbp
+}
+
+func (dbp *DBProvider) SetLogger(logger log.Logger) {
+	dbp.logger = logger.With("label", dbp.label)
+}
+
+func (dbp *DBProvider) SetLimit(limit int) *DBProvider {
+	dbp.limit = limit
+	return dbp
+}
+
+// Implements PersistentProvider.
+func (dbp *DBProvider) SaveFullCommit(fc FullCommit) error {
+
+	dbp.logger.Info("DBProvider.SaveFullCommit()...", "fc", fc)
+	batch := dbp.db.NewBatch()
+
+	// Save the fc.validators.
+	// We might be overwriting what we already have, but
+	// it makes the logic easier for now.
+	vsKey := validatorSetKey(fc.ChainID(), fc.Height())
+	vsBz, err := dbp.cdc.MarshalBinary(fc.Validators)
+	if err != nil {
+		return err
+	}
+	batch.Set(vsKey, vsBz)
+
+	// Save the fc.NextValidators.
+	nvsKey := validatorSetKey(fc.ChainID(), fc.Height()+1)
+	nvsBz, err := dbp.cdc.MarshalBinary(fc.NextValidators)
+	if err != nil {
+		return err
+	}
+	batch.Set(nvsKey, nvsBz)
+
+	// Save the fc.SignedHeader
+	shKey := signedHeaderKey(fc.ChainID(), fc.Height())
+	shBz, err := dbp.cdc.MarshalBinary(fc.SignedHeader)
+	if err != nil {
+		return err
+	}
+	batch.Set(shKey, shBz)
+
+	// And write sync.
+	batch.WriteSync()
+
+	// Garbage collect.
+	// TODO: optimize later.
+	if dbp.limit > 0 {
+		dbp.deleteAfterN(fc.ChainID(), dbp.limit)
+	}
+
+	return nil
+}
+
+// Implements Provider.
+func (dbp *DBProvider) LatestFullCommit(chainID string, minHeight, maxHeight int64) (
+	FullCommit, error) {
+
+	dbp.logger.Info("DBProvider.LatestFullCommit()...",
```

Comment:
I think this should go after `if maxHeight == 0 {`, not before.

Reply:
It can stay here if we want to log what was asked for.

Reply:
I think I prefer knowing what args were passed in...